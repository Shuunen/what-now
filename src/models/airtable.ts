
export interface AirtableTaskRecord {
  id: string;
  createdTime: string;
  fields: {
    name: string;
    once: string;
    'completed-on': string;
    order: number;
    'average-time': number;
  };
}

export interface AirtableResponse {
  records: AirtableTaskRecord[];
  error?: {
    type: string;
    message: string;
  };
}
