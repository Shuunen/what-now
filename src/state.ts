/* c8 ignore next */
import { createState, storage } from 'shuutils'
import type { AirtableTask } from './utils/airtable'

const tasks: AirtableTask[] = []

export const { state, watchState } = createState({
  apiBase: '',
  apiKey: '',
  isLoading: false,
  isSetup: false,
  statusError: '',
  statusInfo: 'Loading, please wait...',
  statusProgress: '',
  tasks,
  tasksTimestamp: 0,
}, storage, ['apiBase', 'apiKey', 'tasks', 'tasksTimestamp'])
