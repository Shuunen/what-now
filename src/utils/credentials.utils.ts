import type { CredentialField } from './state.utils'

/**
 * Parse clipboard
 * @param clipboard the clipboard content
 * @returns the parsed clipboard
 */
export function parseClipboard (clipboard: string) {
  // clipboard can contains something like : "appABC
  // patXYZ.123
  // https://zob.com"
  const regex = /"(?<apiBase>app\w+)\n(?<apiToken>pat[\w.]+)\n(?<hueEndpoint>http[^"]+)"/u
  // eslint-disable-next-line no-useless-assignment
  const { apiBase = '', apiToken = '', hueEndpoint = '' } = regex.exec(clipboard)?.groups ?? {}
  return { apiBase, apiToken, hueEndpoint } satisfies Record<CredentialField, string>
}
