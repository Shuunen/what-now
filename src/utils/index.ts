export * from './date'
export * from './dom'
export * from './event'
export * from './object'
export * from './storage'

