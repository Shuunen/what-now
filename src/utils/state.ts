import { createState, storage } from 'shuutils'
import type { AirtableTask } from '../types'

const tasks: AirtableTask[] = []

export const { state, watchState } = createState({
  apiBase: '',
  apiToken: '',
  hueEndpoint: '',
  isLoading: false,
  isSetup: false,
  showErrorToast: '',
  showInfoToast: '',
  statusError: '',
  statusInfo: 'Loading, please wait...',
  statusProgress: '',
  tasks,
  tasksTimestamp: 0,
}, storage, /* c8 ignore next */['apiBase', 'apiToken', 'hueEndpoint', 'tasks', 'tasksTimestamp'])
