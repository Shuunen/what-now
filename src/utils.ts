import { dom } from 'shuutils'
import { FormField } from './models'

export const button = (content: string, classes = ''): HTMLElement => {
  return dom('button', `bg-blue-800 m-auto sm:ml-0 px-4 py-1 ${classes}`, content)
}

export const form = (fields: FormField[], validate = 'Send form'): HTMLFormElement => {
  const element = dom('form', 'gap-4 grid mt-4') as HTMLFormElement
  element.innerHTML = fields.map(field => `<label class="grid gap-4 sm:grid-cols-3">
    <span>${field.label}</span>
    <input class="bg-blue-900 px-2" name="${field.name}" pattern="${field.pattern}" maxlength="17" required>
    <a class="ml-auto sm:ml-4" href="${field.href}" target="_blank">
      <span class="border-b">${field.link}</span>
      <svg class="h-4 inline ml-2 w-4"><use xlink:href="icons.svg#external"></use></svg>
    </a>
  </label>`).join('')
  element.append(button(validate))
  return element
}

const headersJson = { 'Accept': 'application/json', 'Content-Type': 'application/json' }
export const patch = async (url: string, data: Record<string, unknown>): Promise<Response> => fetch(url, { headers: headersJson, method: 'patch', body: JSON.stringify(data) })
