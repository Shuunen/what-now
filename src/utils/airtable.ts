import type { AirtableResponse } from '../types'
import { logger } from './logger'
import { state } from './state'

const baseLength = 17
const minTokenLength = 50

export function airtableHeaders (token: string) {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return { 'Accept': 'application/json', 'Authorization': `Bearer ${token}`, 'Content-Type': 'application/json' }
}

/* c8 ignore next 11 */
export async function airtablePatch (url: string, data: { [key: string]: unknown }) {
  if (typeof window === 'undefined') return { records: [] }
  const response = await fetch(url, {
    body: JSON.stringify(data),
    headers: airtableHeaders(state.apiToken),
    method: 'PATCH',
  })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return await response.json() as AirtableResponse
}

/* c8 ignore next 9 */
export async function airtableGet (url: string) {
  if (typeof window === 'undefined') return { records: [] }
  const response = await fetch(url, {
    headers: airtableHeaders(state.apiToken),
  })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return await response.json() as AirtableResponse
}

export function airtableValidate (base?: string, token?: string) {
  const isBaseOk = base !== undefined && typeof base === 'string' && base.length === baseLength
  const isTokenOk = token !== undefined && typeof token === 'string' && token.length > minTokenLength
  return isBaseOk && isTokenOk
}

export function airtableUrl (base: string, target: string) {
  return `https://api.airtable.com/v0/${base}/${target}?view=todo`
}

export function checkCredentials (hash = '') {
  logger.info('check credentials', hash.length > 0 ? `and detected hash "${hash}"` : '')
  const matches = /#(?<app>app\w{14})&(?<token>pat[\w.]{50,100})/u.exec(hash)
  if (matches?.groups?.app !== undefined && matches.groups.token !== undefined && airtableValidate(matches.groups.app, matches.groups.token)) {
    state.apiBase = matches.groups.app
    state.apiToken = matches.groups.token
    logger.info('credentials found in hash')
  }
  state.isSetup = airtableValidate(state.apiBase, state.apiToken)
  logger.info('credentials are', state.isSetup ? 'valid' : 'invalid')
  state.statusInfo = state.isSetup ? '' : 'Welcome dear user !'
  return state.isSetup
}

