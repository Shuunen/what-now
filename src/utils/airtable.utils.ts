import type { AirtableResponse } from '../types'
import { logger } from './logger.utils'
import { state } from './state.utils'

const baseLength = 17
const minTokenLength = 50

/**
 * Airtable headers
 * @param token the token to use
 * @returns the headers for airtable
 */
export function airtableHeaders (token: string) {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return { 'Accept': 'application/json', 'Authorization': `Bearer ${token}`, 'Content-Type': 'application/json' }
}

/* c8 ignore next 16 */
/**
 * Airtable patch
 * @param url the url to patch
 * @param data the data to patch
 * @returns the response
 */
export async function airtablePatch (url: string, data: Readonly<Record<string, unknown>>) {
  if (typeof window === 'undefined') return { records: [] }
  const response = await fetch(url, {
    body: JSON.stringify(data),
    headers: airtableHeaders(state.apiToken),
    method: 'PATCH',
  })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return await response.json() as AirtableResponse
}

/* c8 ignore next 14 */
/**
 * Airtable get
 * @param url the url to get
 * @returns the response
 */
export async function airtableGet (url: string) {
  if (typeof window === 'undefined') return { records: [] }
  const response = await fetch(url, {
    headers: airtableHeaders(state.apiToken),
  })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return await response.json() as AirtableResponse
}

/**
 * Airtable validate
 * @param base the base to use
 * @param token the token to use
 * @returns true if the base and token are valid
 */
export function airtableValidate (base?: string, token?: string) {
  const isBaseOk = base !== undefined && typeof base === 'string' && base.length === baseLength
  const isTokenOk = token !== undefined && typeof token === 'string' && token.length > minTokenLength
  return isBaseOk && isTokenOk
}

/**
 * Airtable url
 * @param base the base
 * @param target the target
 * @returns the url
 */
export function airtableUrl (base: string, target: string) {
  return `https://api.airtable.com/v0/${base}/${target}?view=todo`
}

/**
 * Check credentials
 * @param hash the hash to check
 * @returns true if the credentials are valid
 */
// eslint-disable-next-line complexity
export function checkCredentials (hash = '') {
  logger.info('check credentials', hash.length > 0 ? `and detected hash "${hash}"` : '')
  const matches = /#(?<app>app\w{14})&(?<token>pat[\w.]{50,100})/u.exec(hash)
  if (matches?.groups?.app !== undefined && matches.groups.token !== undefined && airtableValidate(matches.groups.app, matches.groups.token)) {
    state.apiBase = matches.groups.app
    state.apiToken = matches.groups.token
    logger.info('credentials found in hash')
  }
  state.isSetup = airtableValidate(state.apiBase, state.apiToken)
  logger.info('credentials are', state.isSetup ? 'valid' : 'invalid')
  state.statusInfo = state.isSetup ? '' : 'Welcome dear user !'
  return state.isSetup
}

