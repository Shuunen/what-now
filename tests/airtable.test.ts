import { expect, it } from 'vitest'
import { airtableValidate, checkCredentials } from '../src/utils/airtable'
import { state } from '../src/utils/state'

it('airtableValidate A', () => {
  expect(airtableValidate('A')).toBe(false)
})

it('checkCredentials A nothing setup', () => {
  expect(checkCredentials()).toBe(false)
})

it('checkCredentials B giving valid hash', () => {
  expect(checkCredentials('#app12345654987123&key12345654987123')).toBe(true)
})

it('checkCredentials C giving invalid hash', () => {
  state.apiBase = ''
  state.apiKey = ''
  const hasSucceed = checkCredentials('#app12343&key12345654987123')
  expect(hasSucceed).toBe(false)
})

it('checkCredentials D with a with valid base & key in state', () => {
  state.apiBase = 'app12345654987123'
  state.apiKey = 'key12345654987123'
  const hasSucceed = checkCredentials()
  expect(hasSucceed).toBe(true)
})
