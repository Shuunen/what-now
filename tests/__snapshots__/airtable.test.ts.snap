// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`airtableHeaders A 1`] = `
{
  "Accept": "application/json",
  "Authorization": "Bearer pat123546",
  "Content-Type": "application/json",
}
`;
